sdk/python/toolchain/uv.go

```diff
@@ -0,0 +1,265 @@
+		if err != nil {
+			if !errors.Is(err, os.ErrNotExist) {
+				return nil, fmt.Errorf("error while looking for pyproject.toml in %s: %w", root, err)
+			}
+			// No uv.lock, do we have a pyproject.toml?
+			pyprojectTomlDir, err := searchup(root, "pyproject.toml")
+			if err != nil {
+				if !errors.Is(err, os.ErrNotExist) {
+					return nil, fmt.Errorf("error while looking for pyproject.toml in %s: %w", root, err)
+				}
+				// We have no uv.lock and no pyproject.toml, place the virtualenv in the project root.
+				virtualenv = filepath.Join(root, defaultVirtualEnv)
+			} else {
+				// We have a pyproject.toml, place the virtualenv next to it.
+				virtualenv = filepath.Join(pyprojectTomlDir, defaultVirtualEnv)
+			}
+		} else {
+			// We have a uv.lock, place the virtualenv next to it.
+			virtualenv = filepath.Join(uvLockDir, defaultVirtualEnv)
+		}
+	}
+	if !filepath.IsAbs(virtualenv) {
+		virtualenv = filepath.Join(root, virtualenv)
+	}
+
+	u := &uv{
+		virtualenvPath: virtualenv,
+		root:           root,
+	}
+
+	// Validate the version
+	cmd := u.uvCommand(context.Background(), "", false, nil, nil, "--version")
+	versionString, err := cmd.Output()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get uv version: %w", err)
+	}
+	version, err := u.uvVersion(string(versionString))
+	if err != nil {
+		return nil, err
+	}
+	logging.V(9).Infof("Python toolchain: using uv version %s", version)
+
+	return u, nil
+}
+
+func (u *uv) InstallDependencies(ctx context.Context, cwd string, useLanguageVersionTools,
+	showOutput bool, infoWriter, errorWriter io.Writer,
+) error {
+	if useLanguageVersionTools {
+		if err := installPython(ctx, cwd, showOutput, infoWriter, errorWriter); err != nil {
+			return err
+		}
+	}
+
+	// Look for a uv.lock file.
+	// If no uv.lock file is found, look for a pyproject.toml file.
+	// If no pyproject.toml file is found, create it, and then look for a
+	// requirements.txt file to install dependencies.
+	if _, err := searchup(cwd, "uv.lock"); err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return fmt.Errorf("error while looking for uv.lock in %s: %w", cwd, err)
+		}
+
+		// No uv.lock found, look for pyproject.toml.
+		if _, err := searchup(cwd, "pyproject.toml"); err != nil {
+			if !errors.Is(err, os.ErrNotExist) {
+				return fmt.Errorf("error while looking for pyproject.toml in %s: %w", cwd, err)
+			}
+
+			// No pyproject.toml found, create one using `uv`
+			initCmd := u.uvCommand(ctx, cwd, showOutput, infoWriter, errorWriter,
+				"init", "--no-readme", "--no-package", "--no-pin-python")
+			if err := initCmd.Run(); err != nil {
+				return fmt.Errorf("error initializing python project: %w", err)
+			}
+
+			// `uv init` creates a `hello.py` file, delete it.
+			contract.IgnoreError(os.Remove(filepath.Join(cwd, "hello.py")))
+
+			// Look for a requirements.txt file and use it to install dependencies
+			requirementsTxtDir, err := searchup(cwd, "requirements.txt")
```

Comment:
If we find a `requirements.txt`, would it make sense to do the `uv init` in the directory of the `requirements.txt`?  That feels a little more natural to me than having it wherever `pulumi` is run potentially.

Reply:
@tgummerer I'm guessing that it is relevant. TL;DR

When migrating the Pulumi Python environment from pip to uv, we will have a `requirements.txt` file. We must migrate what we already have in requirements.txt to `project.toml` to support UV. How to do it?

1. Start UV project:
```
uv init --no-pin-python --no-readme --no-package
```
2. Remove the hello.py file(template):
```
rm hello.py
```
3. Migrate from requirements.txt to pyproject.toml
```
uv add $(cat requirements.txt)
```
Based on that process, I'm guessing that we can improve the code above to support this behaviour before deleting or renaming the `requirements.txt` file.

Reply:
As a suggestion, we can keep the `requirements.txt` file renamed for `.requirements.txt` when we do not delete the user file but remove it from the project scope.

Reply:
Maybe the best thing to do is do nothing?

Reply:
Hrm, I feel like my comment here was misunderstood. https://github.com/pulumi/pulumi/pull/17609#discussion_r1829457779 was about keeping or removing `requirements.txt` *after* we set up UV.

What I'm suggesting here is that if we run pulumi in `/home/user/mypulumiprogram/subdirectory`, but we find a `requirements.txt` at `/home/user/mypulumiprogram`, we currently still initialize `uv` in the `subdirectory`.  I'm wondering if we should instead run `uv init` in `/home/user/mypulumiprogram`, so all the uv data is created at the same directory level as where the `requirements.txt` file lives.

Reply:
@tgummerer , I understood your point. However, we want to solve all corner cases regarding the migration of current Pulumi projects. But what will happen if we use `pulumi new` with a UV template? In that case, we will not have `uv.lock`, `pyproject.toml` and `requirements.txt` because the project has started. No migration is involved. It will be required to run `uv init` to allow us to have a new `pyproject.toml` and `Pulumi.yaml` with new UV toolchain. I suggest keeping it as is and mitigating that behaviour with documentation to help people avoid such mistakes and run the code in the wrong path.  What do you think?

Reply:
If there's no preexisting files, I think our only option is to initialize in the current directory.  My comment was only about what we should do if we do find a `requirements.txt` file.

Reply:
@tgummerer, please correct me if I forgot something, but to summarize what we have:

* If we don't find the `requirements.txt` file, we start the project without migrating dependencies. This behaviour covers new projects;

* If we find the `requirements.txt` file, we start the project by installing the libraries in the new format and keeping requirements.txt as it is. I agree that deleting or renaming may not be a good idea;

What would be the next steps? The code presents a good start and covers the two main processes.

Reply:
I might be missing something, but this comment thread originally was about where we should place `pyproject.toml`.

I implemented the suggestion to use the location of requirements.txt (fallback cwd).